diff --git a/packages/sx05re/emuelec/bin/batocera/emuelec-bluetooth b/packages/sx05re/emuelec/bin/batocera/emuelec-bluetooth
index 0558516f11a..5b2bfcbee0b 100644
--- a/packages/sx05re/emuelec/bin/batocera/emuelec-bluetooth
+++ b/packages/sx05re/emuelec/bin/batocera/emuelec-bluetooth
@@ -142,6 +142,8 @@ class BluetoothCTL:
 
     @pairable.setter
     def agent(self, value):
+        if value != "off":
+            ShellIO.execute(["bluetoothctl", "agent", value])
         ShellIO.execute(["bluetoothctl", "agent", "on" if value else "off"])
 
     def scan(self, timeout=10):
@@ -308,6 +310,10 @@ if __name__ == "__main__":
     if len(sys.argv) >= 2:
         SCAN_TIME = int(sys.argv[1])
 
+    BT_AGENT_TYPE = "NoInputNoOutput"
+    if len(sys.argv) >= 3:
+        BT_AGENT_TYPE = sys.argv[2]
+
     print("running scan for " + str(SCAN_TIME) + " seconds.")
     bt = BluetoothCTL()
 
@@ -320,8 +326,8 @@ if __name__ == "__main__":
         exit()
 
     bt.power = True
-    bt.agent = True
     bt.discoverable = True
+    bt.agent = BT_AGENT_TYPE
     bt.pairable = True
 
     start_time = time.perf_counter()
