diff --git a/cheevos/cheevos.c b/cheevos/cheevos.c
index ace1600552..1d1e33053b 100644
--- a/cheevos/cheevos.c
+++ b/cheevos/cheevos.c
@@ -290,6 +290,12 @@ static void rcheevos_award_achievement(const rc_client_achievement_t* cheevo)
    if (!cheevo)
       return;
 
+	 char pixelcade[256];
+	 snprintf(pixelcade, sizeof(pixelcade), "/storage/.emulationstation/scripts/achievements/pixelcade.sh %u \"%s\" \"%s\" > /dev/null 2>/dev/null",
+	 cheevo->id, cheevo->title, cheevo->description);
+	 //CHEEVOS_LOG(pixelcade);
+	 system(pixelcade);
+			 
    /* Show the on screen message. */
    if (settings->bools.cheevos_visibility_unlock)
    {