diff --git a/core/linux/common.cpp b/core/linux/common.cpp
index 5927b471980..d3fde398a3d 100644
--- a/core/linux/common.cpp
+++ b/core/linux/common.cpp
@@ -67,6 +67,12 @@ void fault_handler(int sn, siginfo_t * si, void *segfault_ctx)
 		return;
 	}
 #endif
+
+	long int i1 = (long int)((void *)ctx.pc);
+	long int i2 = (long int) si->si_addr;
+	if (i1 == i2 && i1 == 0)
+		return;
+
 	ERROR_LOG(COMMON, "SIGSEGV @ %p -> %p was not in vram, dynacode:%d", (void *)ctx.pc, si->si_addr, dyna_cde);
 #ifdef __SWITCH__
 	MemoryInfo meminfo;